### PR TITLE
Add serialize binary method. Add VarInt Utils for serialize & deserializing

### PR DESCRIPTION
**Pull request by vigneshtdev**
Will add deserialize methods, test-cases in a future commit under this PR. 
Need some opinion on use of bytebuffer though.

Addresses #326
Can you look if this https://github.com/dcodeIO/ByteBuffer.js/ package can be used?

### CODE REVIEW

**Comment by vigneshtdev** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
>  **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fcensus-instrumentation%2Fopencensus-node%2Fpull%2F362) for more info**.

I signed it!

File: packages/opencensus-core/src/internal/util.ts [hunk 1]
```diff
@@ -33,3 +33,42 @@ const spanRandomBuffer = randomFillSync ?
 export function randomSpanId() {
   return spanRandomBuffer().toString('hex');
 }
+
+/** Common methods to encode and decode varints and varlongs into ByteBuffers and arrays. */
+export class VarInt {
+  
+  static varIntSize(i: number): number {
```
Comment (draffensperger) [inline packages/opencensus-core/src/internal/util.ts]: Nit: what would you think about making these top-level exported functions rather static members of a `VarInt` class (unless `VarInt` will contain state, in which case they could be member functions)?

File: packages/opencensus-core/src/internal/util.ts [hunk 2]
```diff
@@ -33,3 +33,44 @@ const spanRandomBuffer = randomFillSync ?
...
+/**
+ * Common methods to encode and decode varints and varlongs into ByteBuffers
+ * and arrays.
+ */
+export class VarInt {
+  static varIntSize(i: number): number {
+    let result = 0;
+    do {
+      result++;
+      i >>>= 7;
```
Comment (draffensperger) [inline packages/opencensus-core/src/internal/util.ts]: Could you add comments for what this and the other algorithms below are doing? And maybe also link to an authoritative source for them?

Also, were these algorithms designed for 64-bit ints and do we know they work OK with doubles (JS `number`)? Maybe be sure to test this for very large numbers? (You may be able to use Big Int depending on your Node version target, or a 64-bit int polyfill)

File: packages/opencensus-core/src/internal/util.ts [hunk 3]
```diff
@@ -33,3 +33,44 @@ const spanRandomBuffer = randomFillSync ?
...
+    return offset;
+  }
+
+  static getVarInt(src: Int8Array, offset: number, dst: number[]): number {
+    let result = 0;
+    let shift = 0;
+    let b;
+    do {
+      if (shift >= 32) {
+        throw new Error('varint too lond');
```
Comment (draffensperger) [inline packages/opencensus-core/src/internal/util.ts]: Nit: long not lond

**Comment by songy23** [comment]
Ping @googlebot

File: packages/opencensus-core/src/internal/util.ts (hunk 1 above)
Comment (vigneshtdev) [reply packages/opencensus-core/src/internal/util.ts]: We are making use of bunch of methods like getVarInt, putVarInt,varIntSize so I felt housing them under a class would be the way to go.

**Comment by vigneshtdev** [comment]
```
lerna ERR!   1) Stackdriver Trace Exporter
lerna ERR!        publish()
lerna ERR!          should fail exporting with wrong projectId:
lerna ERR! 
lerna ERR!       AssertionError [ERR_ASSERTION]: false == true
lerna ERR!       + expected - actual
lerna ERR! 
lerna ERR!       -false
lerna ERR!       +true
lerna ERR!       
lerna ERR!       at failExporter.publish.then.result (build/test/test-stackdriver-cloudtrace.js:15:20132)
lerna ERR!       at propagateAslWrapper (/home/node/project/packages/opencensus-core/node_modules/async-listener/index.js:502:23)
lerna ERR!       at /home/node/project/packages/opencensus-core/node_modules/async-listener/index.js:539:70
lerna ERR!       at process._tickCallback (internal/process/next_tick.js:68:7)
lerna ERR! 
l ```

A test in opencensus-propagation-stackdriver is failing for some reason

File: packages/opencensus-core/src/tags/propagation/binary-serializer.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+// import {TextDecoder} from 'util';
+
+import {VarInt} from '../../internal/util';
+import {TagMap} from '../tag-map';
+import {TagKey, TagValue} from '../types';
+
+export let VERSION_ID = 0;
+export let TAG_FIELD_ID = 0;
+// This size limit only applies to the bytes representing tag keys and values.
+const TAG_MAP_SERIALIZED_SIZE_LIMIT = 0;
```
Comment (songy23) [inline packages/opencensus-core/src/tags/propagation/binary-serializer.ts]: Why is this 0? In Java it's [8192](https://github.com/census-instrumentation/opencensus-java/blob/5a97afdefdb2d9f4213264d91fa67b646adbf5d9/impl_core/src/main/java/io/opencensus/implcore/tags/propagation/SerializationUtils.java#L72).

File: packages/opencensus-core/src/tags/propagation/binary-serializer.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+function putVarInt(input: number, byteArray: Int8Array) {
+  const output: Int8Array = new Int8Array(VarInt.varIntSize(input));
+  VarInt.putVarInt(input, output, 0);
+  byteArray = new Int8Array([...byteArray, ...output]);
+}
+
+// function decodeString(byteArray: Int8Array): string {
+//   return new TextDecoder('utf-8').decode(byteArray);
+// }
```
Comment (songy23) [inline packages/opencensus-core/src/tags/propagation/binary-serializer.ts]: nit: new line at EOF.

**Comment by mayurkale22** [comment]
> ```
> lerna ERR!   1) Stackdriver Trace Exporter
> lerna ERR!        publish()
> lerna ERR!          should fail exporting with wrong projectId:
> lerna ERR! 
> lerna ERR!       AssertionError [ERR_ASSERTION]: false == true
> lerna ERR!       + expected - actual
> lerna ERR! 
> lerna ERR!       -false
> lerna ERR!       +true
> lerna ERR!       
> lerna ERR!       at failExporter.publish.then.result (build/test/test-stackdriver-cloudtrace.js:15:20132)
> lerna ERR!       at propagateAslWrapper (/home/node/project/packages/opencensus-core/node_modules/async-listener/index.js:502:23)
> lerna ERR!       at /home/node/project/packages/opencensus-core/node_modules/async-listener/index.js:539:70
> lerna ERR!       at process._tickCallback (internal/process/next_tick.js:68:7)
> lerna ERR! 
> l ```
> 
> A test in opencensus-propagation-stackdriver is failing for some reason
> ```

I am on it.

File: packages/opencensus-core/src/internal/util.ts (hunk 1 above)
Comment (draffensperger) [reply packages/opencensus-core/src/internal/util.ts]: OK. What would you think about using a module for that? There are some static classes used as containers in the library, but I tend to agree with this [Stack Overflow answer](https://stackoverflow.com/a/32984233) that TS modules are a nice way to do that type of grouping vs. static classes. Modules have similar encapsulation and naming (through the file) that a class has but also make it easier to reference (you can destructure if you want).

What do you think?

If you feel strongly about the static methods, I'm open assuming @mayurkale22 is OK.

Comment (vigneshtdev) [reply packages/opencensus-core/src/internal/util.ts]: tslint gives the following error when trying to use modules.
`'namespace' and 'module' are disallowed `

Comment (draffensperger) [reply packages/opencensus-core/src/internal/util.ts]: Oh, I should have specified, I meant just an ES6 module that is created by just making a new `.ts` file like `varint-util.ts` or similar. Could that work?

Comment (vigneshtdev) [reply packages/opencensus-core/src/internal/util.ts]: My first thought was thought, but in the SO question namespace was mentioned, so I assumed that you specifically wanted that. my bad!

File: packages/opencensus-core/src/internal/VarInt.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright 2018 Google LLC
```
Comment (songy23) [inline packages/opencensus-core/src/internal/VarInt.ts]: nit: s/2018/2019

Comment (songy23) [reply packages/opencensus-core/src/internal/VarInt.ts]: and s/Google LLC/OpenCensus Authors

**Comment by draffensperger** [comment]
Thanks for putting this in a new file. Could we name it with `small-caps-and-dashes.ts` to match the other files?

Regarding the desire to add https://github.com/dcodeIO/ByteBuffer.js/, what would be the value added of doing that? My tendency is to avoid a dependency if we don't need it, but may not know the use case here.

File: packages/opencensus-core/src/tags/propagation/binary-serializer.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// import {TextDecoder} from 'util';
+import * as VarInt from '../../internal/VarInt';
+import {TagMap} from '../tag-map';
+import {TagKey, TagValue} from '../types';
+
+export let VERSION_ID = 0;
```
Comment (mayurkale22) [inline packages/opencensus-core/src/tags/propagation/binary-serializer.ts]: ```let``` to ```const```? Same for next line.

File: packages/opencensus-core/src/tags/propagation/binary-serializer.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+// Serializes a TagMap to the on-the-wire format.
+export function serializeBinary(tagMap: TagMap): Int8Array {
+  let totalCharacters = 0;
+  const arr: Int8Array = new Int8Array([VERSION_ID]);
+  const tags = tagMap.tags;
+  if (tags) {
+    tags.forEach((value: TagValue, key: TagKey) => {
+      totalCharacters += key.name.length;
+      totalCharacters += value.value.toString().length;
```
Comment (mayurkale22) [inline packages/opencensus-core/src/tags/propagation/binary-serializer.ts]: No need to use ```toString()``` on ```value```, which is already a string type. Also, value can be a null.

File: packages/opencensus-core/src/tags/propagation/binary-serializer.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+export let VERSION_ID = 0;
+export let TAG_FIELD_ID = 0;
+// This size limit only applies to the bytes representing tag keys and values.
+const TAG_MAP_SERIALIZED_SIZE_LIMIT = 8192;
+
+// Serializes a TagMap to the on-the-wire format.
+export function serializeBinary(tagMap: TagMap): Int8Array {
+  let totalCharacters = 0;
+  const arr: Int8Array = new Int8Array([VERSION_ID]);
```
Comment (mayurkale22) [inline packages/opencensus-core/src/tags/propagation/binary-serializer.ts]: The Node gRPC library expects a ```Buffer``` for binary headers if the normalized key ends with '-bin', in this case our key name is [**"grpc-tags-bin"**](https://github.com/grpc/grpc-java/blob/9a38dea91c6194f7a882e2be1c2846567e9cdcda/core/src/main/java/io/grpc/internal/CensusStatsModule.java#L116). I think we should use ```Buffer``` here. Similar to https://github.com/census-instrumentation/opencensus-node/blob/master/packages/opencensus-propagation-binaryformat/src/binary-format.ts#L53

Comment (vigneshtdev) [reply packages/opencensus-core/src/tags/propagation/binary-serializer.ts]: What should the size be allocated initially? The link given here [max-size](https://github.com/census-instrumentation/opencensus-specs/blob/master/encodings/BinaryEncoding.md#tag-context)
broken.

**Comment by vigneshtdev** [comment]
Still, the same tests are failing. Didn't do git pull before committing though.

**Comment by mayurkale22** [comment]
> Still, the same tests are failing. Didn't do git pull before committing though.

Please rebase, test failure is already fixed in earlier PR.

**Comment by mayurkale22** [comment]
@vigneshtdev Friendly ping.

**Comment by mayurkale22** [comment]
@vigneshtdev We are looking forward to completing this ask asap. Let's know if you are still actively working on this.

**Comment by vigneshtdev** [comment]
Yeah I will post a PR soon. I was looking into making use of  bytebuffer instead int8array

**Comment by mayurkale22** [comment]
@vigneshtdev we have a target deadline to complete this, please let me know if you need any help to complete this PR.  I am open to take over if you are not actively working on this.

**Comment by vigneshtdev** [comment]
@mayurkale22  Sorry for the delays. Feel free to take over this PR.

**Comment by mayurkale22** [comment]
> @mayurkale22 Sorry for the delays. Feel free to take over this PR.

I have raised a PR to complete this task #431, feel free to add reviews there.

**Comment by mayurkale22** [comment]
@vigneshtdev Thanks for the help! I have completed this work in another PR https://github.com/census-instrumentation/opencensus-node/pull/431

**Pull request closed without merging**